### PR TITLE
Rubygem

### PR DESCRIPTION
**Pull request by neha170**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [ ] The toolchain has been rebuilt successfully (or no changes were made to it)
- [ ] The toolchain/worker package manifests are up-to-date
- [ ] Any updated packages successfully build (or no packages were changed)
- [ ] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [ ] All package sources are available
- [ ] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [ ] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [ ] All source files have up-to-date hashes in the `*.signatures.json` files
- [ ] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [ ] Documentation has been updated to match any changes to the build system
- [ ] Ready to merge

---

###### Summary 

What does the PR accomplish, why was it needed?

###### Change Log  

- Change
- Change
- Change

###### Does this affect the toolchain?  

**YES/NO**

###### Associated issues  

- #xxxx

###### Links to CVEs  
- https://nvd.nist.gov/vuln/detail/CVE-YYYY-XXXX

###### Test Methodology

- Pipeline build id: xxxx

### CODE REVIEW

**Review by mfrw — COMMENTED** [review]
I modified the title of the PR. Please feel free to change to something more informative.
I think we need to have a consistent way of commit messages and PR titles :)

A few opinionated resources:

[Linux kernel commit best practices](https://www.kernel.org/doc/html/v4.12/process/submitting-patches.html) [I personally follow it]
[How to write a good commit message](https://cbea.ms/git-commit/)
tl;dr
A simple rule that removes a lot of confusion:

"If applied this commit will: subsystem/spec/package: subject of my commit message"

PS: None of the above is the *right* way; its just, among others, one of the ways :)

File: SPECS/ruby/ruby.spec [hunk 1]
```diff
@@ -190,7 +286,57 @@ sudo -u test make test TESTS="-v"
...
+%{gem_dir}/gems/test-unit-%{test_unit_version}/sample
+%{gem_dir}/gems/test-unit-%{test_unit_version}/test
+%{gem_dir}/specifications/test-unit-%{test_unit_version}.gemspec
+%doc %{gem_dir}/gems/test-unit-%{test_unit_version}/README.md
+%{gem_dir}/gems/test-unit-%{test_unit_version}/Rakefile
+%doc %{gem_dir}/gems/test-unit-%{test_unit_version}/doc
+
 %changelog
+* Tue Mar 15 2022 Neha Agarwal <nehaagarwal@microsoft.com> - 2.7.4-2
+- Build bigdecimal, minitest, test-unit, rdoc and power_assert rubygems (taken from Fedora)
```
Comment (PawelWMS) [inline SPECS/ruby/ruby.spec:338]: Please also mention the version of Fedora and the license under which they release their spec (I'm guessing it's MIT).

File: SPECS/rubygem-bundler/rubygem-bundler.spec [hunk 2]
```diff
@@ -3,14 +3,15 @@
...
+Release:        1%{?dist}
 License:        MIT
 Vendor:         Microsoft Corporation
 Distribution:   Mariner
 Group:          Development/Languages
-URL:            https://rubygems.org/gems/%{gem_name}/versions/%{version}
-Source0:        https://rubygems.org/downloads/%{gem_name}-%{version}.gem
+URL:            https://github.com/rubygems/rubygems
+#Source0:        https://github.com/rubygems/rubygems/archive/refs/tags/bundler-v%{version}.tar.gz
+Source0:        rubygems-%{gem_name}-v%{version}.tar.gz
```
Comment (PawelWMS) [inline SPECS/rubygem-bundler/rubygem-bundler.spec:14]: This is preferred now (makes automation/PR checks easier):

```suggestion
Source0:        https://github.com/rubygems/rubygems/archive/refs/tags/bundler-v%{version}.tar.gz#/rubygems-%{gem_name}-v%{version}.tar.gz
```

File: SPECS-EXTENDED/rubygem-abrt/rubygem-abrt.spec [hunk 3]
```diff
@@ -1,29 +1,24 @@
...
+Summary:       ABRT support for Ruby
+Name:          rubygem-%{gem_name}
+Version:       0.4.0
+Release:       2%{?dist}
+License:       MIT
+URL:           https://github.com/voxik/abrt-ruby
+Vendor:        Microsoft Corporation
+Distribution:  Mariner
+#Source0:      https://github.com/voxik/abrt-ruby/archive/refs/tags/v%{version}.tar.gz
+Source0:       %{gem_name}-ruby-%{version}.tar.gz
```
Comment (PawelWMS) [inline SPECS-EXTENDED/rubygem-abrt/rubygem-abrt.spec:11]: Preferred:

```suggestion
Source0:      https://github.com/voxik/abrt-ruby/archive/refs/tags/v%{version}.tar.gz#/%{gem_name}-ruby-%{version}.tar.gz
```

May I ask you to update other modified specs in the same way?

File: SPECS-EXTENDED/rubygem-abrt/rubygem-abrt.spec [hunk 4]
```diff
@@ -1,29 +1,24 @@
...
-URL: http://github.com/voxik/abrt-ruby
-Source0: https://rubygems.org/gems/%{gem_name}-%{version}.gem
-# git clone https://github.com/voxik/abrt-ruby.git && cd abrt-ruby
-# git archive -v -o abrt-0.4.0-spec.tar.gz v0.4.0 spec/
-Source1: %{gem_name}-%{version}-spec.tar.gz
-Requires: libreport-filesystem
+Summary:       ABRT support for Ruby
+Name:          rubygem-%{gem_name}
+Version:       0.4.0
+Release:       2%{?dist}
```
Comment (PawelWMS) [inline SPECS-EXTENDED/rubygem-abrt/rubygem-abrt.spec:5]: Please update the `Release` number and changelog entry. This is a substantial change from what we had.

File: SPECS-EXTENDED/rubygem-asciidoctor/rubygem-asciidoctor.spec [hunk 5]
```diff
@@ -1,35 +1,63 @@
...
+Version:        2.0.17
 Release:        1%{?dist}
 Group:          Development/Languages
 License:        MIT
 Vendor:         Microsoft Corporation
 Distribution:	Mariner
-URL:            https://rubygems.org/gems/%{gem_name}/versions/%{version}
-Source0:        https://rubygems.org/downloads/%{gem_name}-%{version}.gem
+URL:            https://asciidoctor.org/
+Source0:        https://github.com/%{gem_name}/%{gem_name}/archive/refs/tags/v%{version}.tar.gz#%{gem_name}-%{version}.tar.gz
```
Comment (PawelWMS) [inline SPECS-EXTENDED/rubygem-asciidoctor/rubygem-asciidoctor.spec:13]: Are we sure this works? Usually there's a `/` needed in front of the file name:

```suggestion
Source0:        https://github.com/%{gem_name}/%{gem_name}/archive/refs/tags/v%{version}.tar.gz#/%{gem_name}-%{version}.tar.gz
```

File: SPECS-EXTENDED/rubygem-coderay/rubygem-coderay.spec [hunk 6]
```diff
@@ -35,53 +30,44 @@ BuildArch: noarch
...
-%exclude %{gem_cache}
-%{gem_spec}
+%exclude /%{gemdir}/cache
+/%{gemdir}/specifications
 
 %files doc
-%doc %{gem_docdir}
+%doc /%{gemdir}/doc/
 %doc %{gem_instdir}/README_INDEX.rdoc
+/%{gemdir}/gems/%{gem_name}-%{version}.rc1/README_INDEX.rdoc
```
Comment (PawelWMS) [inline SPECS-EXTENDED/rubygem-coderay/rubygem-coderay.spec:65]: This looks suspicious. If it's a documentation file, then let's add it with the `%doc` macro. But it also looks similar to the file one line above it. Are we sure this is needed?

File: SPECS-EXTENDED/rubygem-coderay/rubygem-coderay.spec [hunk 7]
```diff
@@ -35,53 +30,44 @@ BuildArch: noarch
...
 
 %changelog
 * Fri Oct 15 2021 Pawel Winogrodzki <pawelwi@microsoft.com> - 1.1.2-8
 - Initial CBL-Mariner import from Fedora 32 (license: MIT).
+- License verified.
```
Comment (PawelWMS) [inline SPECS-EXTENDED/rubygem-coderay/rubygem-coderay.spec:70]: Please create update the `Release` tag and add the changelog entries there. Especially since I have most certainly not verified the license.;)

File: SPECS-EXTENDED/rubygem-console/rubygem-console.spec [hunk 8]
```diff
@@ -9,8 +9,9 @@ License:        MIT
 Vendor:         Microsoft Corporation
 Distribution:   Mariner
 Group:          Development/Languages
-URL:            https://rubygems.org/gems/%{gem_name}/versions/%{version}
-Source0:        https://rubygems.org/downloads/%{gem_name}-%{version}.gem
+URL:            https://socketry.github.io/console/
+#Source0:        https://github.com/socketry/console/archive/refs/tags/v%{version}.tar.gz
+Source0:        %{gem_name}-%{version}.tar.gz
```
Comment (PawelWMS) [inline SPECS-EXTENDED/rubygem-console/rubygem-console.spec:14]: Don't forget to update the `Release` tag!

File: SPECS-EXTENDED/rubygem-deep_merge/rubygem-deep_merge.spec [hunk 9]
```diff
@@ -8,29 +8,30 @@ Version:        1.2.1
 Release:        1%{?dist}
 Group:          Development/Languages
 License:        MIT
-Vendor:		Microsoft Corporation
+Vendor:         Microsoft Corporation
 Distribution:	Mariner
-URL:            https://rubygems.org/gems/%{gem_name}/versions/%{version}
-Source0:        https://rubygems.org/downloads/%{gem_name}-%{version}.gem
+URL:            https://github.com/danielsdeleo/deep_merge
```
Comment (PawelWMS) [inline SPECS-EXTENDED/rubygem-deep_merge/rubygem-deep_merge.spec:13]: Last one, I promise: please update the `Release` tag and changelog where that's not done, yet.

**Review by PawelWMS — CHANGES_REQUESTED** [review]
A couple of blockers, mostly repeated ones, nothing big so far. Please see comments for more details.

File: SPECS-EXTENDED/rubygem-flexmock/rubygem-flexmock.spec [hunk 10]
```diff
@@ -60,36 +62,26 @@ popd
...
 %{gem_spec}
 
-%files	doc
-%{gem_instdir}/todo.txt
-%{gem_instdir}/doc/
+%files doc
 %{gem_docdir}/
 
 %changelog
-* Thu Feb 24 2022 Pawel Winogrodzki <pawelwi@microsoft.com> - 2.3.6-8
```
Comment (PawelWMS) [inline SPECS-EXTENDED/rubygem-flexmock/rubygem-flexmock.spec:88]: Please don't remove changelog entries.;) Especially since now the latest entry's release number doesn't match the spec's `Release` tag value.

File: SPECS-EXTENDED/rubygem-ronn/rubygem-ronn.spec [hunk 11]
```diff
@@ -10,16 +11,16 @@ Summary:        Manual authoring tool
...
+Source0:        %{gem_name}-%{version}.tar.gz
 BuildArch:      noarch
 BuildRequires:  rubygems-devel
-Requires:       rubygem(hpricot) rubygem(rdiscount) rubygem(mustache) rubygems groff-base
-
-%if 0%{?fedora} >= 19 || 0%{?rhel} > 7
-Requires: ruby(release)
-%else
-Requires: ruby(abi) >= 1.9.1
-%endif
```
Comment (PawelWMS) [inline SPECS-EXTENDED/rubygem-ronn/rubygem-ronn.spec:22]: Thanks for the clean-up!

File: SPECS-EXTENDED/rubygem-semantic_puppet/rubygem-semantic_puppet.spec [hunk 12]
```diff
@@ -8,30 +8,31 @@ Version:        1.0.4
...
 %build
+gem build %{gem_name}
 
 %install
-gem install -V --local --force --install-dir %{buildroot}/%{gemdir} %{SOURCE0}
+gem install -V --local --force --install-dir %{buildroot}/%{gemdir} %{gem_name}-%{version}.gem
 
 %files
 %defattr(-,root,root,-)
-%license %{gemdir}/gems/%{gem_name}-%{version}/LICENSE
```
Comment (PawelWMS) [inline SPECS-EXTENDED/rubygem-semantic_puppet/rubygem-semantic_puppet.spec:31]: Is there no license information at all in the new source now? We'd need something. If there's none, we should add a `Source` file explaining how we know the license for this package and include that files through the `%license` macro.

Look for "LICENSE.PTR" in other specs to see examples.

Comment (neha170) [reply SPECS-EXTENDED/rubygem-semantic_puppet/rubygem-semantic_puppet.spec:31]: Added license file back from source.

File: SPECS-EXTENDED/rubygem-abrt/rubygem-abrt.spec [hunk 13]
```diff
@@ -1,29 +1,23 @@
...
-Requires: libreport-filesystem
+Summary:       ABRT support for Ruby
+Name:          rubygem-%{gem_name}
+Version:       0.4.0
+Release:       3%{?dist}
+License:       MIT
+URL:           https://github.com/voxik/abrt-ruby
+Vendor:        Microsoft Corporation
+Distribution:  Mariner
+Source0:      https://github.com/voxik/abrt-ruby/archive/refs/tags/v%{version}.tar.gz#/%{gem_name}-ruby-%{version}.tar.gz
```
Comment (PawelWMS) [inline SPECS-EXTENDED/rubygem-abrt/rubygem-abrt.spec:10]: Nit:
```suggestion
Source0:       https://github.com/voxik/abrt-ruby/archive/refs/tags/v%{version}.tar.gz#/%{gem_name}-ruby-%{version}.tar.gz
```

File: SPECS-EXTENDED/rubygem-hpricot/rubygem-hpricot.spec [hunk 14]
```diff
@@ -9,20 +7,16 @@
...
 # ext/fast_xs/FastXsService.java is licensed under ASL 2.0
 License:	MIT AND ASL 2.0
 Vendor:	        Microsoft Corporation
 Distribution:	Mariner
 URL:		http://github.com/hpricot/hpricot
-# Non-free file removed, see Source10
-# Source0:	http://gems.rubyforge.org/gems/%{gem_name}-%{version}.gem
-Source0:	https://src.fedoraproject.org/repo/pkgs/rubygem-hpricot/hpricot-0.8.6-modified.gem/5d737234720f3847cc5def6c0047f2e4/hpricot-0.8.6-modified.gem
-Source10:	rubygem-hpricot-create-free-gem.sh
+Source0:    https://github.com/hpricot/hpricot/archive/refs/tags/%{version}.tar.gz#/%{gem_name}-%{version}.tar.gz
```
Comment (PawelWMS) [inline SPECS-EXTENDED/rubygem-hpricot/rubygem-hpricot.spec:16]: Nit:
```suggestion
Source0:        s://github.com/hpricot/hpricot/archive/refs/tags/%{version}.tar.gz#/%{gem_name}-%{version}.tar.gz
```

File: SPECS-EXTENDED/rubygem-pkg-config/rubygem-pkg-config.spec [hunk 15]
```diff
@@ -1,42 +1,29 @@
...
 Summary:	A pkg-config implementation by Ruby
 Name:		rubygem-%{gem_name}
 Version:	1.4.5
-Release:	3%{?dist}
+Release:	4%{?dist}
 License:	LGPLv2+
-URL:		http://github.com/rcairo/pkg-config
-
-Source0:	http://rubygems.org/gems/%{gem_name}-%{version}.gem
+URL:	    http://github.com/rcairo/pkg-config
```
Comment (PawelWMS) [inline SPECS-EXTENDED/rubygem-pkg-config/rubygem-pkg-config.spec:13]: Nit:
```suggestion
URL:	         http://github.com/rcairo/pkg-config
```

File: SPECS-EXTENDED/rubygem-rake-compiler/rubygem-rake-compiler.spec [hunk 16]
```diff
@@ -1,18 +1,14 @@
...
+Release:	1%{?dist}
 License:	MIT
 Vendor:		Microsoft Corporation
 Distribution:	Mariner
-URL:		http://rake-compiler.rubyforge.org/
-Source0:	https://rubygems.org/gems/%{gem_name}-%{version}.gem
-
-Requires:	ruby(release)
+URL:		https://rubydoc.info/gems/rake-compiler
+Source0:    https://github.com/rake-compiler/rake-compiler/archive/refs/tags/v%{version}.tar.gz#/%{gem_name}-%{version}.tar.gz
```
Comment (PawelWMS) [inline SPECS-EXTENDED/rubygem-rake-compiler/rubygem-rake-compiler.spec:11]:
```suggestion
Source0:        https://github.com/rake-compiler/rake-compiler/archive/refs/tags/v%{version}.tar.gz#/%{gem_name}-%{version}.tar.gz
```

**Review by PawelWMS — APPROVED** [review]
Looks good, thank you for the amazing work! My remaining comments are minor suggestions regarding whitespaces, no blockers.
